plugins/gstperf.c

```diff
@@ -198,13 +235,73 @@ gst_perf_get_property (GObject * object, guint property_id,
   }
 }
 
+static gboolean
+gst_perf_update_bps (void *data)
+{
+  guint buffer_current_idx;
+  GstPerf *perf;
+
+  g_return_val_if_fail(data, FALSE);
+
+  perf = GST_PERF (data);
+
+  /* Calculate bits per second */
+  perf->bps =
+      perf->byte_count * GST_PERF_BITS_PER_BYTE / (perf->bps_frequency / 1000.0);
+
+  g_mutex_lock (&perf->bps_mutex);
+  perf->byte_count = G_GUINT64_CONSTANT (0);
+  g_mutex_unlock (&perf->bps_mutex);
+
+  /* Update bps average */
+  if (!perf->bps_window_size) {
+    perf->mean_bps =
+        gst_perf_update_average (perf->byte_count_total, perf->bps,
+        perf->mean_bps);
+    perf->byte_count_total++;
+  } else {
+    /*
+     * Moving average uses a circular buffer, get index for next value which
+     * is the oldest sample, this is the same as the value were the new sample
+     * is to be stored
+     */
+    buffer_current_idx = (perf->byte_count_total) % perf->bps_window_size;
+
+    perf->mean_bps =
+        gst_perf_update_moving_average (perf->bps_window_size, perf->mean_bps,
+        perf->bps, perf->bps_window_buffer[buffer_current_idx]);
+
+    perf->bps_window_buffer[buffer_current_idx] = perf->bps;
```

Comment:
This is not a moving average filter. You are implementing a strange recursive filter. You have:

```
y(n) = 1/3 * (y(n-1) + y(n-2) + x(n))
```

while you should have:

```
y(n) = 1/3 (x(n) + x(n-1) + x(n-2))
```